### ISSUE TITLE
move packedReis to db, add delay server boolean to server_config.json

### ISSUE THREAD

**Issue by nief** [irods/irods#3049]
- [x] master
- [x] 4-2-stable

------

hello,
      in iRODS architecture, only one irodsReServer must run per zone. It is usually running on the iCAT enabled server, so no problem (default setting). However, in a high availability setup, one has to chose one of the server to host the rule engine server. In v<4.x, there was a "reHost" parameter in server.config.
In 4.1.7, it does not appear to be available for setup in server_config.json. I did not see anything in the doc as well.
This is an essential feature for people having a high availability architecture.
cheers,
JY

**Comment by trel** [comment]
Every catalog service provider in a Zone is currently running the irodsReServer.  This can lead to race conditions with long-executing rules.

By adding the `reHost` config, this would allow a server to delegate the rule execution server duties to another machine on startup.

We would also need to have the python control scripts be capable of spinning up *only* the irodsReServer on a machine.

**Comment by trel** [comment]
After further discussion, it seems that rather than re-introducing the `reHost` variable, we can solve this more completely with the following two pieces of work:

1. Move the storage of the `packedRei`s (which each hold an environment for a delayed rule) from `/var/lib/irods/config/packedRei/` into the database.  We are proposing these `packedRei` values live in AVUs affiliated with the Zone itself.  This will solve the single source of truth problem for where this information should live when there is more than one catalog provider in a Zone.  We are interested in solving this elegantly since upcoming use cases will demand having many of the servers in a Zone (if not all of them) be catalog providers.

2. Introduce a new 'rule_execution_server' key into the `server_config.json` schema.  This key would be a boolean and determine whether an `irodsReServer` would be launched on each particular server.  It will be well-documented that if there are more than one of these set to `true` in a single Zone, you may experience race conditions when servicing the delayed execution queue.  We will expect there to be exactly one server set to `true` in each Zone.  Facilities for checking and/or healing this setting across servers could be included later, but will not be included at this time.

It will not be necessary to have the capability of launching _only_ an `irodsReServer` on a particular machine.

The two settings will afford a server to be in one of four possible states:

<table>
<tr><td></td><td></td><td colspan='2'>rule_execution_server</td></tr>
<tr><td></td><td></td><td>true</td><td>false</td></tr>
<tr><td rowspan='2'>catalog_service_role</td><td>provider</td><td>normal, expected for first/only catalog provider</td><td>for catalog provider 2-n</td></tr>
<tr><td>consumer</td><td>possible, but not likely</td><td>normal, expected</td></tr>
</table>

@nief, would you please weigh in on this proposal.  Can you think of a use case where this will not handle what you would like to do?

**Comment by nief** [comment]
Thanks.
Point 1 makes perfect sense, it will remove a dependency on a local file system. As these informations will be stored into the database, would it solve the possible race conditions if several irodsReServer were running (maybe some locks would need to be added) ? In the future, that would be good to be able to run at least one more irodsReServer in a given zone. It is not strongly needed but it would be a plus. We have been living with that for years: if the server which is hosting irodsReServer is down for several hours, the delayed rules are just processed later. However, under these circumstances, we can end up with a lot of backlog (such as delayed replication rules, if clients continue to put files into iRODS during the outage).
As for point 2, I am ok with that. How will it discover the server on which irodsReServer is running as we would use a boolean variable ? 
thanks,
JY

**Comment by trel** [comment]
Putting information into the database would not solve the race condition problem.  The race condition occurs because a particular `irodsReServer` wakes up, asks for what work it needs to do, gets a list, and then works through it with _n_ `irodsAgent`s concurrently on different processors on a single machine (currently set by `maximum_number_of_concurrent_rule_engine_server_processes`) (renaming that for clarity is https://github.com/irods/irods/issues/3399).  Having multiple machines running `irodsReServer`s would mean that a single rule on the delayed queue could get picked up and worked on by more than one `irodsReServer`... which is the problem.

We would need to implement a worker queue of some kind (or use RedisQueue, etc.) to have a multiple writer/reader scenario for the delayed rule execution to avoid this race condition.  That could be future work, but is not included in these two proposed bits of work.

To your second question, there is no 'discovery' of the `irodsReServer`.  It is merely the server that wakes up and asks the database for work that is ready to be executed.  The `irodsReServer` knows where the catalog provider is, not the other way around.  Each server's config would have the boolean setting, and it is up to the grid administrator to ensure that there is only one server in the Zone that has `rule_execution_server` set to `true`.

**Comment by trel** [comment]
A boolean is sufficient because the hostname of the irodsReServer will no longer matter.  The host location was included in `reHost` due to the Zone needing to know where to write the `packedRei` files.  This will no longer be necessary because the `packedRei` information will go into the catalog.

**Comment by nief** [comment]
hello,
I have seen that the milestone for this one is 5.0.0. When is it expected to be released ? 
For a HA setup, this feature is mandatory unless one is not using delayed rules.
thanks,
jy

**Comment by trel** [comment]
I agree this should be moved back to 4.3.0.   HA is a valid and critical use case and should be considered as we update the schemas and ready for the next major release.  Thanks for the bump.

**Comment by trel** [comment]
Another request for HA use case in https://github.com/irods/irods/issues/5153.

**Comment by trel** [comment]
The 4.2 codebase has developed a more natural comfort with JSON inputs and data structures in the last couple releases.  I think we can get the `packedRei`s into the database as JSON, and add a new option to the `setup_irods.py` script to control the setting of the new advanced setting in `server_config.json` without needing a configuration schema update.

Proposed advanced setting (rather than `rule_execution_server` above, as we're moving away from that name in 4.3):

```
{
    "advanced_settings": {
        "run_delay_server": true (or false),
}
```

The default values of `run_delay_server`:

<table>
<tr><td>"catalog_service_role": "provider"</td><td>"run_delay_server": true</td></tr>
<tr><td>"catalog_service_role": "consumer"</td><td>"run_delay_server": false</td></tr>
</table>


If set, this new option will override the default values above:
`python /var/lib/irods/scripts/setup_irods.py --run_delay_server true`
or
`python /var/lib/irods/scripts/setup_irods.py --run_delay_server false`

This plan can allow us to make these changes for 4.2.x...

---
Still need to determine/plan the database location for the `packedRei` data.

Consider we add a new column to `r_rule_exec` named `exe_context`?  Is 2700 characters big enough for the JSON?

Or we repurpose... `r_rule_exec.rei_file_path`...

```
ICAT=# \d r_rule_exec;
                Table "public.r_rule_exec"
       Column       |          Type           | Modifiers 
--------------------+-------------------------+-----------
 rule_exec_id       | bigint                  | not null
 rule_name          | character varying(2700) | not null
 rei_file_path      | character varying(2700) | 
 user_name          | character varying(250)  | 
 exe_address        | character varying(250)  | 
 exe_time           | character varying(32)   | 
 exe_frequency      | character varying(250)  | 
 priority           | character varying(32)   | 
 estimated_exe_time | character varying(32)   | 
 notification_addr  | character varying(250)  | 
 last_exe_time      | character varying(32)   | 
 exe_status         | character varying(32)   | 
 create_ts          | character varying(32)   | 
 modify_ts          | character varying(32)   | 
Indexes:
    "idx_rule_exec" UNIQUE, btree (rule_exec_id)
```

**Comment by kript** [comment]
A counter proposal is to have the server that is to run rules be in the catalog; one a server starts up it checks the catalog to see if its the server that should run the rules engine. On a new zone the first provider spun up adds itself, but this could be amended later.
This could be done with a new iadmin command e.g.`iadmin reServer name <XXX>`.

My issues with using `server_config.json` is that the change will need to be made on every server in the Zone and maintained and could be forgotten, especially if the configuration is not maintained/deployed by automation.
In addition, having an iadmin command to switch the rules server over would allow an HA process such as corosync to switch the rules engine in the event of a failure, rather than needing to edit a config file or switch between files.

**Comment by trel** [comment]
We would provide a default behavior (provider -> true, consumer -> false) in the case where the `server_config.json` entry is not present - so a change will not be required on every server in the Zone.  This is a first step towards the 4.3.0 work in https://github.com/irods/irods/issues/3993.

Will need to think through your counter proposal of storing the host/location of the Zone's delay server in the catalog.  This would require a bit more machinery/signaling between servers to coordinate any 'switching' within a live Zone.

**Comment by jasoncoposky** [comment]
I would hesitate to place configuration into the catalog as it breaks the Single Point of Truth principle.  I agree that a strong default of 'I do not run rules' if the flag is not present makes sense, and any changes in the configuration should be also handled by the upgrade scripts.

**Comment by kript** [comment]
I am probably missing something, but the catalog *is* the single point of truth isn't it? Whereas each server having its own `server_config.json` has a separate source of truth as to whether it is the rule/delay server for the zone?
I think my suggestion is no good if there are intended to be more than one rules/delay server, but I thought that was the restriction?

**Comment by jasoncoposky** [comment]
@kript My point is that the SPoT for configuration of a given server is /etc/irods/server_config.json.  If we start splitting that between the configuration file and the catalog we have now violated the SPoT principle.  I feel that a server should not need to reach out to the catalog to determine if it should start the rule execution server also as that is very early in the boot process of the system.  The server does not have a connection to the catalog until an agent is spawned by a client connection.

**Comment by kript** [comment]
>  My point is that the SPoT for configuration of a given server is /etc/irods/server_config.json. If we start splitting that between the configuration file and the catalog we have now violated the SPoT principle.

Yes, however given #4145 show is this has already happened and has no (announced) plan for fixing...

> The server does not have a connection to the catalog until an agent is spawned by a client connection.

Fair enough.

**Comment by trel** [comment]
I think the following distinction may be helpful...

This is a discussion about whether this setting (and perhaps others like it) are server-specific settings, or Zone-wide settings.

1) server-specific -> single point of truth is in each server's `server_config.json`
2) Zone-wide -> single point of truth is in the Zone's catalog/database

The current thinking is that whether a machine should be running an instance of the `irodsReServer`/`irodsDelayServer` is a server-specific setting (option 1).  Each server "independently" stands up an `irodsServer` based on its configuration and then services incoming requests accordingly.  It does not depend on conversation/coordination with other servers in the Zone, including the catalog provider.

The point about 'live-switching' the location of the `irodsDelayServer` within a Zone is an interesting one - that is a definite vote for option 2.  A background thread in each `irodsServer` could regularly ping (every 30min or so, by default) the catalog provider to find out if it should be running an `irodsDelayServer` on its own host (or more generally, receive any Zone-wide configurations it may need to know about).  If `yes`, then it spins one up.  If `no`, then it could gracefully finish any jobs in flight and stop its running `irodsDelayServer`.  Could be a part of https://github.com/irods/irods/issues/4977?

**Comment by trel** [comment]
We are planning to run with option 1 for now - each `server_config.json` will determine whether to start an instance of the  `irodsReServer`/`irodsDelayServer`.

'Switching' the location of the delay server within a Zone will be an exercise in updating two machine's `server_config.json` files: turning the original one 'off', and turning a different one 'on'.

---

@nief There should be no need for running multiple delay servers - we will document that this could result in race conditions.

From a comment above (https://github.com/irods/irods/issues/3049#issuecomment-339422551)... (slightly updated for nomenclature)...

> Introduce a new 'run_delay_server' key into the `server_config.json` schema. This key would be a boolean and determine whether an `irodsReServer`/`irodsDelayServer` would be launched on each particular server. It will be well-documented that if there are more than one of these set to `true` in a single Zone, you may experience race conditions when servicing the delayed execution queue. We will expect there to be exactly one server set to `true` in each Zone. Facilities for checking and/or healing this setting across servers could be included later, but will not be included at this time.

To address your comment about throttling... the delay server is much more performant than in the past.   And rules can always run their work via `remote()` to execute on a separate machine.  Adding an implicit `remote()` (so that jobs get automatically spread around the Zone) is https://github.com/irods/irods/issues/4429.

**Comment by trel** [comment]
@korydraughn we need to make sure we 'migrate' any existing `packedRei`s on disk into the database on upgrade.

**Comment by trel** [comment]
The current plan...

As a consolidated update to... https://github.com/irods/irods/issues/3049#issuecomment-339422551

1) Move the storage of the `packedRei`s (which each hold an environment for a delayed rule) from `/var/lib/irods/config/packedRei/` into the database as a JSON string. This information should move to a new column `r_rule_exec.exe_context varchar(2700)`. Any `dataObjInfo_t` replica list information should be changed to just the relevant `data_id`, making the size of the stored information deterministic (and relatively small).  We will fetch the replica list upon use, if necessary.

2) On upgrade, introduce a new boolean 'run_delay_server' key into the `server_config.json` schema. This key will determine whether an `irodsReServer`/`irodsDelayServer` would be launched on each particular server. It will be well-documented that if there are more than one of these set to `true` in a single Zone, you may experience race conditions when servicing the delayed execution queue. We will expect there to be exactly one server set to `true` in each Zone. Facilities for checking and/or healing this setting across servers could be included later, but will not be included at this time.  Servers without the new `run_delay_server` key defined will have a default value of `false`.

3) On upgrade, automatically migrate any existing on-disk `packedRei`s into the new database column (by matching on the existing path name in `r_rule_exec.rei_file_path`).  The `r_rule_exec.rei_file_path` column will now be unused moving forward.

4) Introduce a new option `--run_delay_server` to the `setup_irods.py` script for new installations.  This value will override the defaults for a new provider (`true`) or new consumer (`false`).

**Comment by kript** [comment]
This is such a helpful thread, thank you all. A few comments and questions, if I may;

>'Switching' the location of the delay server within a Zone will be an exercise in updating two machine's server_config.json files: turning the original one 'off', and turning a different one 'on'.

Can you document what you expect to happen if a rule is supposed to fire and there are no rule servers to service it? Also the best practice, i.e. should you turn all servers off before enabling the new one? I'm  thinking of what might be needed for designing the HA actions to migrate the role around.

> To address your comment about throttling... the delay server is much more performant than in the past.

If you have any stats here that would be very interesting! Also any best practice on how to spec a server based on the expected throughput of delay rules (I appreciate this can be tricky without knowing what the rules may do!).

> we need to make sure we 'migrate' any existing packedReis on disk into the database on upgrade.
You have probably already thought of this, but when there are multiple providers to be upgraded, is there a way to determine which should be upgraded first? Slightly worried about the 'phantom' packedReis being uploaded as 'good', or if there are multiple with the race condition aluded to earlier, are multiple servers allowed to upload to the table? Item 3 might already be aimed at covering this, of course...

> Move the storage of the packedReis (which each hold an environment for a delayed rule) from /var/lib/irods/config/packedRei/ into the database as a JSON string.

I hear that later versions of some of the supported databases can be made to understand JSON natively as a field, which might be an optimisation win if the database supports it? Might be interesting to explore that and also if the back end database is upgraded, to a version that does now support it, have a way of migrating to the new more optimal format? 

In general with update 1 of the plan, I wonder if a DBA with expertise in each of the supported three databases could be included/sought to suggest optimisations (that could be optional)? We might be able to supply all three of these (but I haven't asked them yet!), if there was the opportunity.

> Facilities for checking and/or healing this setting across servers could be included later, but will not be included at this time.

Any thoughts/notes you might have, especially on healing/fixing, would be welcome as part of the notes, I feel, even if its the database queries etc that might need to be run/files tidied up until more robust tooling is available.

> Introduce a new option --run_delay_server to the setup_irods.py script for new installations. This value will override the defaults for a new provider (true) or new consumer (false).

So via this option you could supply a new true/false? I read this as a toggle to reverse the default, which may not be what you meant (and sounds hard to parse!). 
Also, can this option be included in the setup JSON that can be passed to a server to configure it on first run (we're planning on making more use of this to build servers now were on the giddy heights of 4.2.7)?

**Comment by trel** [comment]
> This is such a helpful thread, thank you all. A few comments and questions, if I may;
> 
> > 'Switching' the location of the delay server within a Zone will be an exercise in updating two machine's server_config.json files: turning the original one 'off', and turning a different one 'on'.
> 
> Can you document what you expect to happen if a rule is supposed to fire and there are no rule servers to service it? Also the best practice, i.e. should you turn all servers off before enabling the new one? I'm thinking of what might be needed for designing the HA actions to migrate the role around.
> 

If there is no delay server to query for, retrieve, and then launch queued delay rules, they'll just sit happily in the database until a delay server *does* come along and ask for the next chunk of jobs to launch.

> > To address your comment about throttling... the delay server is much more performant than in the past.
> 
> If you have any stats here that would be very interesting! Also any best practice on how to spec a server based on the expected throughput of delay rules (I appreciate this can be tricky without knowing what the rules may do!).
> 

I don't know that we have any specific numbers... @alanking?   The earlier implementation was limited by the fetching of 256 jobs per wakeup (by default every 30s) and any long-running job would block the retrieval of other jobs, regardless of the setting for concurrent threads).  Both of those design limitations are now removed.  The delay server now picks up all jobs whose time has come and burns through them as fast as its number of concurrent threads can work.

Related issues here: #3782, #4250, #4251, #4266 (this is what we referred to as the @tedgin curve)

> > we need to make sure we 'migrate' any existing packedReis on disk into the database on upgrade.
> > You have probably already thought of this, but when there are multiple providers to be upgraded, is there a way to determine which should be upgraded first? Slightly worried about the 'phantom' packedReis being uploaded as 'good', or if there are multiple with the race condition aluded to earlier, are multiple servers allowed to upload to the table? Item 3 might already be aimed at covering this, of course...
> 

I think this will not be an issue - as a particular machine is upgraded, we'll check for any `packedRei`s on disk, and just migrate them into the table.  I can't think of any reason this would stomp on itself at the moment - each delayed rule should be unique and have a ready/waiting one-for-one slot in the existing database table.  But we'll keep an eye on an existing HA configuration as a use case.

> > Move the storage of the packedReis (which each hold an environment for a delayed rule) from /var/lib/irods/config/packedRei/ into the database as a JSON string.
> 
> I hear that later versions of some of the supported databases can be made to understand JSON natively as a field, which might be an optimisation win if the database supports it? Might be interesting to explore that and also if the back end database is upgraded, to a version that does now support it, have a way of migrating to the new more optimal format?
> 
> In general with update 1 of the plan, I wonder if a DBA with expertise in each of the supported three databases could be included/sought to suggest optimisations (that could be optional)? We might be able to supply all three of these (but I haven't asked them yet!), if there was the opportunity.
> 

I don't know that the JSON parsing will be that big of a deal to optimize here.  But we're happy to discuss it some more.  It's definitely an open question.

I would guess, at this time, that the parsing of the JSON string will pale in comparison to the work that is about to be done (database lookups, network traffic, reading/writing bits to/from disk) for each delayed rule.

> > Facilities for checking and/or healing this setting across servers could be included later, but will not be included at this time.
> 
> Any thoughts/notes you might have, especially on healing/fixing, would be welcome as part of the notes, I feel, even if its the database queries etc that might need to be run/files tidied up until more robust tooling is available.
> 

Detecting would be pretty easy, as `izonereport` can already gather all the configurations from each server.  It (or another tool) could trivially check for multiple servers set to `true`.  We do not currently have any facilities to reach out and edit the `server_config.json` file on other servers - so that would be a new set of functionality (and security concerns).

No database queries here - just that `run_delay_server` entry in `server_config.json`.

> > Introduce a new option --run_delay_server to the setup_irods.py script for new installations. This value will override the defaults for a new provider (true) or new consumer (false).
> 
> So via this option you could supply a new true/false? I read this as a toggle to reverse the default, which may not be what you meant (and sounds hard to parse!).
> Also, can this option be included in the setup JSON that can be passed to a server to configure it on first run (we're planning on making more use of this to build servers now were on the giddy heights of 4.2.7)?

Yes, on setup of a 4.2.9+ server (but less than 4.3.0), this option would override the default setting of `provider`->`run_delay_server=true` and `consumer`->`run_delay_server=false`.  This would allow HA setups to define this setting correctly from initialization of a new server.

And yes, the unattended setup would not be affected - it would accept this new setting correctly.

And note, 4.3.0+ will change the questions asked by `setup_irods.py` to include this new setting - so the option will no longer be necessary (and can be detected/ignored by `setup_irods.py`).

**Comment by alanking** [comment]
I do not have any official benchmarks for the reworked delayed execution server. Here is a summary of the changes from the UGM 2019 Technology Update: http://slides.com/irods/ugm2019-technology-update#/8 This is slightly out of date as the server has since been refactored to use the `irods::query_processor` but that is just an abstraction for the mechanisms described in the slide, so it is all still true.

When it was first implemented, there was not a cap on the number of connections which could be opened to execute jobs but it is now limited by settings in the server configuration (more on that can be found in this blog post: https://irods.org/2020/04/migrating-between-gpfs-filesystems-with-irods-at-kth/ @ilarik may have some numbers from that deployment).

That said, initial tests were able to chew through ~30,000 rules in about a minute, regardless of the configured sleep time. Even with the corrected limitation of open connections, the old RE server could only execute a maximum of 256 rules per wake-up, being unable to process new delay jobs in the meantime.

Not sure if that answers the question - hope that helps

**Comment by korydraughn** [comment]
Due to #5249  and #5258, the following items will **no longer** be implemented:
- Adding `--run_delay_server=<boolean>` to the setup script
- Adding `run_delay_server` to the server configuration file

The sole purpose of this issue is now about moving the REI files into the catalog.

**Comment by trel** [comment]
oh, still need to revert a couple things.
